solana/programs/swap-layer/src/processor/initiate/transfer.rs

```diff
@@ -0,0 +1,194 @@
+    #[account(mut)]
+    /// The payer of the transaction. This could either be the recipient or a relayer.
+    payer: Signer<'info>,
+
+    custodian: CheckedCustodian<'info>,
+
+    /// CHECK: This account must be the one who paid to create the staged outbound account.
+    #[account(
+        mut,
+        address = staged_outbound.info.prepared_by,
+    )]
+    prepared_by: UncheckedAccount<'info>,
+
+    /// Staging for outbound transfer. This account has all of the instructions needed to initiate
+    /// the transfer.
+    ///
+    /// This account will be closed by the end of the instruction.
+    #[account(
+        mut,
+        close = prepared_by,
+    )]
+    staged_outbound: Account<'info, StagedOutbound>,
+
+    /// This custody token account will be closed by the end of the instruction.
+    #[account(
+        mut,
+        token::mint = common::USDC_MINT,
+        seeds = [
+            crate::STAGED_CUSTODY_TOKEN_SEED_PREFIX,
+            staged_outbound.key().as_ref(),
+        ],
+        bump = staged_outbound.info.custody_token_bump,
+    )]
+    staged_custody_token: Account<'info, token::TokenAccount>,
+
+    /// CHECK: This account must equal the usdc refund token encoded in the staged outbound account.
+    #[account(address = staged_outbound.usdc_refund_token)]
+    usdc_refund_token: UncheckedAccount<'info>,
+
+    /// Peer used to determine whether assets are sent to a valid destination.
+    #[account(
+        constraint = {
+            require_eq!(
+                staged_outbound.info.target_chain,
+                target_peer.seeds.chain,
+                SwapLayerError::InvalidPeer,
+            );
+
+            true
+        }
+    )]
+    target_peer: RegisteredPeer<'info>,
+
+    /// CHECK: Token router config.
+    token_router_custodian: UncheckedAccount<'info>,
+
+    /// CHECK: Mutable, seeds must be \["prepared-order", staged_outbound.key()\].
+    #[account(
+        mut,
+        seeds = [
+            crate::PREPARED_ORDER_SEED_PREFIX,
+            staged_outbound.key().as_ref(),
+        ],
+        bump,
+    )]
+    prepared_order: UncheckedAccount<'info>,
+
+    /// CHECK: Mutable, seeds must be \["prepared-custody", prepared_order.key()\].
+    #[account(mut)]
+    prepared_custody_token: UncheckedAccount<'info>,
+
+    usdc: Usdc<'info>,
+
+    token_router_program: Program<'info, token_router::program::TokenRouter>,
+    token_program: Program<'info, token::Token>,
+    system_program: Program<'info, System>,
+}
+
+pub fn initiate_transfer_new(ctx: Context<InitiateTransferNew>) -> Result<()> {
+    // We perform this operation first so we can have an immutable reference to the staged outbound
+    // account.
+    let staged_redeem = std::mem::take(&mut ctx.accounts.staged_outbound.staged_redeem);
+
+    let staged_outbound = &ctx.accounts.staged_outbound;
+    let custody_token = &ctx.accounts.staged_custody_token;
+
+    // Decode the output token to verify that it's valid.
```

Comment:
Don't think we need this comment anymore, we verify that it's valid when staging.

Reply:
I wanted to keep this here in case of a weird event where the output token encoding changes between staging and the transfer.

This decoding should always succeed. But if there is some staged outbound transfer sitting around for a while and we fire this off to another chain, the transfer might not be able to be completed.

Though the encoding could change while the transfer is in flight, too. So if we add pausing to the Swap Layer program, we could control this even more (and there could be a staged outbound account existing before the program is paused).

How does this all sound?